include/engine/routing_algorithms/routing_base_mld.hpp

```diff
@@ -0,0 +1,283 @@
+                                           forward_heap,
+                                           reverse_heap,
+                                           parent_cell,
+                                           middle,
+                                           weight,
+                                           forward_search_radius,
+                                           reverse_search_radius);
+        }
+        if (!reverse_heap.Empty() && (reverse_heap.MinKey() < reverse_search_radius))
+        {
+            progress = true;
+            routingStep<REVERSE_DIRECTION>(facade,
+                                           partition,
+                                           cells,
+                                           reverse_heap,
+                                           forward_heap,
+                                           parent_cell,
+                                           middle,
+                                           weight,
+                                           reverse_search_radius,
+                                           forward_search_radius);
+        }
+    } while (progress);
+
+    // No path found for both target nodes?
+    if (weight == INVALID_EDGE_WEIGHT || SPECIAL_NODEID == middle)
+    {
+        return std::make_tuple(
+            INVALID_EDGE_WEIGHT, SPECIAL_NODEID, SPECIAL_NODEID, std::vector<EdgeID>());
+    }
+
+    // Get packed path as edges {level, from node ID, to node ID, edge ID}
+    std::vector<std::tuple<LevelID, NodeID, NodeID, EdgeID>> packed_path;
+    NodeID current_node = middle, parent_node = forward_heap.GetData(middle).parent;
+    while (parent_node != current_node)
+    {
+        const auto &data = forward_heap.GetData(current_node);
+        packed_path.push_back(std::make_tuple(data.level, parent_node, current_node, data.edge_id));
+        current_node = parent_node;
+        parent_node = forward_heap.GetData(parent_node).parent;
+    }
+    std::reverse(std::begin(packed_path), std::end(packed_path));
+    const NodeID source_node = current_node;
+
+    current_node = middle, parent_node = reverse_heap.GetData(middle).parent;
+    while (parent_node != current_node)
+    {
+        const auto &data = reverse_heap.GetData(current_node);
+        packed_path.push_back(std::make_tuple(data.level, current_node, parent_node, data.edge_id));
+        current_node = parent_node;
+        parent_node = reverse_heap.GetData(parent_node).parent;
+    }
+    const NodeID target_node = current_node;
+
+    // Unpack path
+    std::vector<EdgeID> unpacked_path;
+    unpacked_path.reserve(packed_path.size());
+    for (auto &packed_edge : packed_path)
+    {
+        LevelID level;
+        NodeID source, target;
+        EdgeID edge_id;
+        std::tie(level, source, target, edge_id) = packed_edge;
+        if (edge_id != SPECIAL_EDGEID)
+        { // a base graph edge
+            unpacked_path.push_back(edge_id);
```

Comment:
I would keep `edge_id` in heap data, because a flag `is_overlay` is  needed to distinguish edge types. Here `is_overlay` flag  is `edge_id == SPECIAL_EDGEID`.